torchtext/modules/multiheadattention.py

```diff
@@ -0,0 +1,203 @@
+        head_dim = k.size(-1) // self.nhead
+        k = k.reshape(src_len, bsz * self.nhead, head_dim)
+
+        assert v.size(-1) % self.nhead == 0, "value's embed_dim must be divisible by the number of heads"
+        head_dim = v.size(-1) // self.nhead
+        v = v.reshape(src_len, bsz * self.nhead, head_dim)
+
+        attn_output, attn_output_weights = self.attention_layer(q, k, v, attn_mask=attn_mask,
+                                                                bias_k=bias_k, bias_v=bias_v)
+        attn_output = attn_output.reshape(tgt_len, bsz, embed_dim)
+        attn_output = self.out_proj(attn_output)
+        return attn_output, attn_output_weights
+
+
+class ScaledDotProduct(torch.nn.Module):
+
+    def __init__(self, dropout=0.0):
+        r"""Processes a projected query and key-value pair to apply
+        scaled dot product attention.
+
+        Args:
+            dropout (float): probability of dropping an attention weight.
+
+        Examples::
+            >>> SDP = torchtext.models.ScaledDotProduct(0.1)
+            >>> q = torch.randn(256, 21, 3)
+            >>> k = v = torch.randn(256, 21, 3)
+            >>> attn_output, attn_weights = SDP(q, k, v)
+            >>> print(attn_output.shape, attn_weights.shape)
+            torch.Size([256, 21, 3]) torch.Size([256, 21, 21])
+        """
+        super(ScaledDotProduct, self).__init__()
+        self.dropout = dropout
+
+    def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor,
+                attn_mask: Optional[torch.Tensor] = None,
+                bias_k: Optional[torch.Tensor] = None,
+                bias_v: Optional[torch.Tensor] = None) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""Uses a scaled dot product with the projected key-value pair to update
+        the projected query.
+
+        Args:
+            query (Tensor): Projected query
+            key (Tensor): Projected key
+            value (Tensor): Projected value
+            attn_mask (BoolTensor, optional): 3D mask that prevents attention to certain positions.
+            bias_k and bias_v:bias: the additional key and value sequence to be added at sequence dim (dim=-3).
+                Those are used for incremental decoding.
+
+        Shape:
+            - query: :math:`(L, N * H, E / H)`
+            - key: :math:`(S, N * H, E / H)`
+            - value: :math:`(S, N * H, E / H)`
+            - attn_mask: :math:`(N * H, L, S)`, positions with ``True`` are not allowed to attend
+                while ``False`` values will be unchanged.
+            - bias_k and bias_v:bias: :math:`(1, N * H, E / H)`
+
+            - Output: :math:`(L, N * H, E / H)`, :math:`(N * H, L, S)`
+
+            where L is the target length, S is the source length, H is the number
+            of attention heads, N is the batch size, and E is the embedding dimension.
+        """
+        if bias_k is not None and bias_v is not None:
```

Comment:
It looks like `bias_k` and `bias_v` are only used if both are defined. If the user provides only one, it is silently ignored. It would be helpful to explicitly document that either both arguments should be provided or neither argument should be provided. Alternatively, you could add an an exception if just one is provided.